File: bayesNet.py
# Arthor: Andrew Lewis
# Project 3: CPSC 352
# April 24th 2019

import random

# conditional probability tables for the Bayesian network as
# a dictionary.  
# 
# The key is a string tuple describing the node.
# Entries are stored as ordered dictionary pairs

bn = {'Course':[[],{'long':0.5, 'short':0.5}],
      'Weather':[[],{'coldWet':0.3, 'hot':0.2, 'nice':0.5}],
      'HarePerf':[['Course','Weather'],
               {('short','coldWet'):{'slow':0.5,'medium':0.3,'fast':0.2},
               ('short','hot'):{'slow':0.1,'medium':0.2,'fast':0.7},
                ('short','nice'):{'slow':0.0,'medium':0.2,'fast':0.8},
                ('long','coldWet'):{'slow':0.7,'medium':0.2,'fast':0.1},
                ('long','hot'):{'slow':0.2,'medium':0.4,'fast':0.4},
                ('long','nice'):{'slow':0.1,'medium':0.3,'fast':0.6}}],
      'TortoisePerf':[['Course', 'Weather'],
                {('short','coldWet'):{'slow':0.2,'medium':0.3,'fast':0.5},
                ('short','hot'):{'slow':0.4,'medium':0.5,'fast':0.1},
                ('short','nice'):{'slow':0.3,'medium':0.5,'fast':0.2},
                ('long','coldWet'):{'slow':0.2,'medium':0.4,'fast':0.4},
                ('long','hot'):{'slow':0.2,'medium':0.5,'fast':0.3},
                ('long','nice'):{'slow':0.4,'medium':0.4,'fast':0.2}}],
      'HareWins':[['HarePerf', 'TortoisePerf'],
                   {('slow','slow'):{'win':0.5},('slow','medium'):{'win':0.1},('slow','fast'):{'win':0.0},
                   ('medium','slow'):{'win':0.8},('medium','medium'):{'win':0.5},('medium','fast'):{'win':0.2},
                   ('fast','slow'):{'win':0.9},('fast','medium'):{'win':0.7},('fast','fast'):{'win':0.5}}]}

# a list of the variables, starting "from the bottom" of the network.
# in topological ordering towards the top
varss = ['Weather','Course','HarePerf','TortoisePerf','HareWins']

# compute probability that var has the value val.  e are the list of
# variable values we already know, and bn has the conditional probability
# tables.
def Pr(var, e, bn):
    parents = bn[var][0]
    parentVals = ()
    # debugprint('Pr***', var, e, bn, parents)
    if len(parents) == 0:
        table = bn[var][1]
        # print(table)
    else:
        # debugprint('   Pr***')
        for parent in parents:
            # print('Var ', e[parent], parent)
            parentVals+=(e[parent])
        table = bn[var][1][parentVals]
        # print(table)
    return table
    
# Generate 
def generateSample(algo, bn, varss, weather = None, course = None):
    # list of generated samples
    e = {}
    if weather is not None:
        e['Weather'] = (weather,)

    if course is not None:
        e['Course'] = (course,)

    for var in varss:
        # pick a value for var according to the probabilities in bn and the
        # ones we already picked.
        if var == 'Weather' and weather is not None:
            continue
        if var == 'Course' and course is not None:
            continue

        r = random.uniform(0.0,1.0)
        probabilities = Pr(var,e,bn)
        lastProb = 0.0
        # print(r)
        for x in probabilities:
            if r <= lastProb+probabilities[x] and r > lastProb:
                e[var] = (x,)
                # print('chosen ', e[var])
                break
            else:
                lastProb+=probabilities[x]
        if 'HareWins' not in e:
            e['HareWins'] = ('lose',)
        
    # return the sample - e
    if algo == 'prior':
        if e['HareWins'] == ('win',):
            return 1
        return 0
    if algo == 'reject':
        return e

# Generates samples and return probability of HareWins
def priorSample(n, weather = None, course = None):
    wins = 0.0
    total = n
    for i in range(n):
        wins+=generateSample('prior',bn, varss, weather, course)    
    if wins is not 0:
        return wins/n
    else:
        return 'No wins'

# Loop through generated samples and filters accordingly
def rejectionSample(n, weather = None, course = None, HarePerf = None, TortoisePerf = None, HareWins = None):
    finalSamples = []
    for i in range(0, n):
        sample = generateSample('reject', bn,varss)
        #print(sample)
        if weather is not None and sample['Weather'] != (weather,):
            # print(weather)
            continue
        if course is not None and sample['Course'] != (course,):
            # print(sample['Course'])
            continue
        if HarePerf is not None and sample['HarePerf'] != (HarePerf,):
            continue
        if TortoisePerf is not None and sample['TortoisePerf'] != (TortoisePerf,):
            continue
        if HareWins is not None and sample['HareWins'] != (HareWins,):
            #print(sample)
            continue
        finalSamples.append(sample) 
        #print(finalSamples)  
    return finalSamples    
   

while True:
    x = input('''"Enter number corresponding to the queries below
    1. In general, how likely is the Hare to win?
    2. Given that is it coldWet, how likely is the Hare to win?
    3. Given that the Tortoise won on the short course, what is the probability distribution for the Weather?"''')
    print("")
    if x == '1' or x == 1: 
        print(priorSample(10000))
        continue
    if x == '2' or x == 2:
        print(priorSample(10000, 'coldWet'))
    if x == '3' or x == 3:
        coldWet = 0.0
        hot = 0.0
        nice = 0.0

        for s in rejectionSample(10000, course = 'short', HareWins='lose'):
            if s['Weather'] == ('coldWet',):
                coldWet+=1.0
            if s['Weather'] == ('hot',):
                hot+=1.0
            if s['Weather'] == ('nice',):
                nice+=1.0

        total = coldWet + hot + nice
        print(total)
        print('coldWet: ', coldWet/total)
        print('Hot: ', hot/total)
        print('Nice: ', nice/total)

File: BayesNet.java
/**
 * Simple class for approximate inference based on the Wet Grass network.
 */
public class BayesNet {
	
	// The list of nodes in the Bayes net
	private Node[] nodes;
	
	// A collection of examples describing whether it is { raining, playing
	// }
	private static final boolean[][] PLAYING_EXAMPLES =

    { { true, false }, { true, false }, { true, true }, 
    { true, false }, { true, false }, { false, true },
    { false, false }, { false, true },{ false, true }, 
    { false, true }, { false, false },{ false, true }, 
    { false, true } };
        
	/**
	 * Constructor that sets up the Wet Grass network.
	 */
	public BayesNet() {
		
		nodes = new Node[6];
		
		nodes[0] = new Node("Course", new Node[] {}, new double[] { 0.4 });
		nodes[1] = new Node("Weather", new Node[] {}, new double[] { 0.002 });
		nodes[2] = new Node("HarePerf", new Node[] { nodes[0], nodes[1] },
							new double[] { 0.02, 0.1, 0.15, 0.5 });
		nodes[3] = new Node("TortoisePerf", new Node[] { nodes[0] }, new double[] {
				0.8, 0.1 });
		nodes[4] = new Node("HareWins", new Node[] { nodes[3] },
							calculatePlayOutsideProbabilities(PLAYING_EXAMPLES));
	}

	/**
	 * Another reason to dislike java.  Why can you not test for membership
	 * in an array?
	 * http://stackoverflow.com/a/12635769
	 */
	public static <T> boolean contains( final T[] array, final T v ) {
		for ( final T e : array )
        if ( e == v || v != null && v.equals( e ) )
            return true;
	
    return false;
	}
	
	/**
	 * Prints the current state of the network to standard out.
	 */
	public void printState() {
		
		for (int i = 0; i < nodes.length; i++) {
			if (i > 0) {
				System.out.print(", ");
			}
			System.out.print(nodes[i].name + " = " + nodes[i].value);
		}
		System.out.println();
	}
	
	/**
	 * Calculates the probability that a child will play outside based on
	 * whether it is raining or not. Returns the conditional probabilities that
	 * a child will play when it is raining and when it is not raining.
	 * 
	 * @param rainingInstances
	 *            A set of training examples in the form { raining, playing }
	 *            from which to compute the probabilities.
	 * @return The probability that a child will play outside when it is {
	 *         raining, !raining }.
	 */
	public double[] calculatePlayOutsideProbabilities(
													  boolean[][] rainingInstances) {
		double[] probabilities = new double[2];
		
		int[] playing = new int[2];
		int[] total = new int[2];
		
		// Count the number of times the child plays for both
		// raining and not raining.
		for (boolean[] sample : rainingInstances) {
		    if (sample[0]) {
				// raining
				// Increment the playing count if the child
				// is playing, and increment the total sample
				// count.
				playing[0] += sample[1] ? 1 : 0;
				total[0]++;
		    } else {
				// not raining
				// Increment the playing count if the child
				// is playing, and increment the total sample
				// count.
				playing[1] += sample[1] ? 1 : 0;
				total[1]++;
		    }
		}
		
		// P(p|r)
		probabilities[0] = (double)playing[0] / (double)total[0];
		// P(p|~r)
		probabilities[1] = (double)playing[1] / (double)total[1];
		
		return probabilities;
	}

	/**
	 * This method will return the probability of a node being true
	 * given its parents.
	 *
	 * @param n The node to test
	 *
	 * @return P(n=true | parents(n))
	 */
	public double probGivenParents(Node n) {
		// TODO: Make this general somehow
		switch (n.parents.length) {
		case 0:
			return n.probs[0];
		case 1:
			return n.parents[0].value ? n.probs[0] : n.probs[1];
		case 2:
			return n.parents[0].value ?
				n.parents[1].value ? n.probs[0] : n.probs[1]
				: n.parents[1].value ? n.probs[1] : n.probs[2];
		default:
			throw new RuntimeException("Only supports up to two parents!");
		}
	}
	
	/**
	 * This method assigns new values to the nodes in the network by sampling
	 * from the joint distribution (based on PRIOR-SAMPLE method from text
	 * book/slides).
	 */
	public void priorSample() {
	    double rand;
	    double prob;
		
	    for (Node n : nodes) {
			prob = probGivenParents(n);
			rand = Math.random();
			n.value = rand <= prob ? true : false;
	    }
	}

	private boolean testModel(int[] indicesOfEvidenceNodes, boolean[] evidenceValues) {
		for (int i = 0; i < indicesOfEvidenceNodes.length; i++) {
			if (nodes[indicesOfEvidenceNodes[i]].value != evidenceValues[i])
				return false;
		}
		
		return true;
	}
	
	/**
	 * Rejection sampling. Returns probability of query variable being true
	 * given the values of the evidence variables, estimated based on the given
	 * total number of samples (see REJECTION-SAMPLING method from text
	 * book/slides).
	 * 
	 * The nodes/variables are specified by their indices in the nodes array.
	 * The array evidenceValues has one value for each index in
	 * indicesOfEvidenceNodes. See also examples in main().
	 * 
	 * @param queryNode
	 *            The variable for which rejection sampling is calculating.
	 * @param indicesOfEvidenceNodes
	 *            The indices of the evidence nodes.
	 * @param evidenceValues
	 *            The values of the indexed evidence nodes.
	 * @param N
	 *            The number of iterations to perform rejection sampling.
	 * @return The probability that the query variable is true given the
	 *         evidence.
	 */
	public double rejectionSampling(int queryNode,
									int[] indicesOfEvidenceNodes, boolean[] evidenceValues, int N) {
		int[] counts = new int[2]; // {true, false} counts for the query node

		// Take the samples
		for (int i = 0; i < N; i++) {
			this.priorSample();
			if (testModel(indicesOfEvidenceNodes, evidenceValues)) {
				if (nodes[queryNode].value) {
					counts[0]++;
				} else {
					counts[1]++;
				}
			}
		}

		return (double)counts[0] / (double)(counts[0] + counts[1]);
	}
	
	/**
	 * MCMC inference. Returns probability of query variable being true given
	 * the values of the evidence variables, estimated based on the given total
	 * number of samples (see MCMC-ASK method from text book/slides).
	 * 
	 * The parameters are the same as in the case of rejectionSampling().
	 * 
	 * @param queryNode
	 *            The variable for which rejection sampling is calculating.
	 * @param indicesOfEvidenceNodes
	 *            The indices of the evidence nodes.
	 * @param evidenceValues
	 *            The values of the indexed evidence nodes.
	 * @param N
	 *            The number of iterations to perform rejection sampling.
	 * @return The probability that the query variable is true given the
	 *         evidence.
	 */
	public double MCMCask(int queryNode, int[] indicesOfEvidenceNodes,
						  boolean[] evidenceValues, int N) {
		
		return 0; // REPLACE THIS LINE BY YOUR CODE
    }
    /**
	 * Inner class for representing a node in the network.
	 */
	private class Node {
		
		// The name of the node
		private String name;
		
		// The parent nodes
		private Node[] parents;
		
		// The probabilities for the CPT
		private double[] probs;
		
		// The current value of the node
		private boolean value;
		
		/**
		 * Initializes the node.
		 */
		private Node(String n, Node[] pa, double[] pr) {
			name = n;
			parents = pa;
			probs = pr;
		}
		
		/**
		 * Returns conditional probability of value "true" for the current node
		 * based on the values of the parent nodes.
		 * 
		 * @return The conditional probability of this node, given its parents.
		 */
		private double conditionalProbability() {
			
			int index = 0;
			for (int i = 0; i < parents.length; i++) {
				if (parents[i].value == false) {
					index += Math.pow(2, parents.length - i - 1);
				}
			}
			return probs[index];
		}
	}
	
	/**
	 * The main method, with some example method calls.
	 */
	public static void main(String[] ops) {
		
		// Create network.
		BayesNet b = new BayesNet();
		
		double[] playOutsideProbabilities = b
			.calculatePlayOutsideProbabilities(PLAYING_EXAMPLES);
		System.out.println("When it is raining, I play outside "
						   + (playOutsideProbabilities[0] * 100) + "% of the time.");
		System.out.println("When it is not raining, I play outside "
						   + (playOutsideProbabilities[1] * 100) + "% of the time.");

		System.out.println();
		
		// Sample five states from joint distribution and print them
		for (int i = 0; i < 5; i++) {
			b.priorSample();
			b.printState();
		}
		
		// Print out results of some example queries based on rejection
		// sampling.
		// Same should be possible with likelihood weighting and MCMC inference.
		System.out.println("\nRejection Sampling");

		// Probability of rain given wet grass and NOT playing outside.
		System.out.print("P(rain | wet grass, ~playing outside) = ");
		System.out.println(b.rejectionSampling(3, new int[] { 4, 5 },
											   new boolean[] { false, true }, 10000));
		
		// Probability of sprinklers given a drought
		System.out.print("P(sprinklers | drought) = ");
		System.out.println(b.rejectionSampling(2, new int[] { 1 },
											   new boolean[] { true }, 100000));
		
		// Probability of wet grass given rain and sprinklers
		System.out.print("P(wet grass | rain, sprinklers) = ");
		System.out.println(b.rejectionSampling(5, new int[] { 2, 3 },
											   new boolean[] { true, true }, 10000000));
	}
}